src/tuner/algo/allreduce/ring.h

```diff
@@ -0,0 +1,311 @@
+ *
+ * @param proto The protocol (NCCL_PROTO_SIMPLE, NCCL_PROTO_LL, or NCCL_PROTO_LL128).
+ * @return The overhead denominator for the given protocol.
+ */
+static inline unsigned long long get_overhead_denominator(int proto)
+{
+    if      (NCCL_PROTO_LL     == proto) { return NCCL_OFI_TUNER_RING_LL_OVERHEAD_DENOMINATOR; }
+    else if (NCCL_PROTO_LL128  == proto) { return NCCL_OFI_TUNER_RING_LL128_OVERHEAD_DENOMINATOR; }
+    else if (NCCL_PROTO_SIMPLE == proto) { return NCCL_OFI_TUNER_RING_SIMPLE_OVERHEAD_DENOMINATOR; }
+    assert(false);
+    __builtin_unreachable();
+}
+
+/**
+ * @brief Get the buffer size for a given protocol.
+ *
+ * @param proto The protocol (NCCL_PROTO_SIMPLE, NCCL_PROTO_LL, or NCCL_PROTO_LL128).
+ * @return The buffer size for the given protocol.
+ */
+static inline unsigned long long get_buffsize(int proto)
+{
+    if      (NCCL_PROTO_LL     == proto) { return NCCL_OFI_TUNER_NCCL_LL_BUFFSIZE; }
+    else if (NCCL_PROTO_LL128  == proto) { return NCCL_OFI_TUNER_NCCL_LL128_BUFFSIZE; }
+    else if (NCCL_PROTO_SIMPLE == proto) { return NCCL_OFI_TUNER_NCCL_BUFFSIZE; }
+    assert(false);
+    __builtin_unreachable();
+}
+
+/**
+ * @brief Get the number of chunk steps for a given protocol.
+ *
+ * @param proto The protocol (NCCL_PROTO_SIMPLE, NCCL_PROTO_LL, or NCCL_PROTO_LL128).
+ * @return The number of chunk steps for the given protocol.
+ */
+static inline unsigned long long get_ring_chunk_steps(int proto)
+{
+    if      (NCCL_PROTO_LL     == proto) { return NCCL_OFI_TUNER_RING_LL_CHUNK_STEPS; }
+    else if (NCCL_PROTO_LL128  == proto) { return NCCL_OFI_TUNER_RING_LL128_CHUNK_STEPS; }
+    else if (NCCL_PROTO_SIMPLE == proto) { return NCCL_OFI_TUNER_RING_SIMPLE_CHUNK_STEPS; }
+    assert(false);
+    __builtin_unreachable();
+}
+
+/**
+ * @brief Get the step size for a given protocol.
+ *
+ * @param proto The protocol (NCCL_PROTO_SIMPLE, NCCL_PROTO_LL, or NCCL_PROTO_LL128).
+ * @return The step size for the given protocol.
+ */
+static inline unsigned long long get_ring_stepsize(int proto)
+{
+    return get_buffsize(proto) / NCCL_OFI_TUNER_NCCL_STEPS;
+}
+
+/**
+ * @brief Get the chunk size for a given protocol.
+ *
+ * @param proto The protocol (NCCL_PROTO_SIMPLE, NCCL_PROTO_LL, or NCCL_PROTO_LL128).
+ * @return The step size for the given protocol.
+ */
+static inline unsigned long long get_ring_chunksize(int proto)
+{
+    return (get_ring_stepsize(proto) * get_ring_chunk_steps(proto) / get_overhead_denominator(proto)) * get_overhead_numerator(proto);
+}
+
+/**
+ * @brief Get the chunk window in bytes.
+ *
+ * @param proto The protocol (NCCL_PROTO_SIMPLE, NCCL_PROTO_LL, or NCCL_PROTO_LL128).
+ * @return The number of chunks for the given protocol.
+ */
+static inline unsigned long long get_ring_chunk_window_in_bytes(int proto)
+{
+	return get_ring_chunk_steps(proto)*get_buffsize(proto)*get_overhead_numerator(proto)/get_overhead_denominator(proto)/NCCL_OFI_TUNER_NCCL_STEPS;
```

Comment:
In NCCL code, `/get_overhead_denominator(proto)` is done first, then followed by `get_overhead_numerator(proto)'. 
For example, for LL128, it is `(x / 16) * 15` instead of `(x * 15) / 16`.

In this file, `get_ring_chunksize` already correctly follows NCCL code (see code line 212).

Can we do `return get_ring_chunksize(proto)`?

Reply:
Made this change and it did not impact results. I previously had overflow issues, and copied this carelessly to try to figure it out. Clearly that's been handled, replacing it with the old version had no impact on the previous results. Thanks.